# Atwell - Homework 1 Notebook 3

##Packages, Setup, and Data Import

In [1]:
%matplotlib inline

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn.model_selection  import train_test_split, cross_val_score, KFold
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, r2_score, explained_variance_score
from sklearn.metrics import roc_curve, auc, confusion_matrix, classification_report, mean_squared_error, median_absolute_error

####Check working directory and change if needed.

In [3]:
%pwd

'C:\\Users\\mlatw\\Desktop\\BIA6303\\Assignment_1'

In [4]:
cd /Users/mlatw/Desktop/BIA6303/Assignment_1

C:\Users\mlatw\Desktop\BIA6303\Assignment_1


Importing the community crime data set and displaying some basic information about it. The list of column datatypes is long, but necessary to see what obstacles we face before conducting data mining.

In [5]:
pd.options.display.max_rows = 999 #added to display all rows (if needed, but output should be tweaked as required)
url_data="https://archive.ics.uci.edu/ml/machine-learning-databases/communities/communities.data"
comm_crime = pd.read_csv(url_data, na_values="?") #NA values in this dataset are encoded with a ?
comm_crime.columns = [#'crimepredict'
'state','county','community','communityname','fold','population','householdsize','racepctblack','racePctWhite','racePctAsian','racePctHisp','agePct12t21','agePct12t29','agePct16t24','agePct65up','numbUrban','pctUrban','medIncome','pctWWage','pctWFarmSelf','pctWInvInc','pctWSocSec','pctWPubAsst','pctWRetire','medFamInc','perCapInc','whitePerCap','blackPerCap','indianPerCap','AsianPerCap','OtherPerCap','HispPerCap','NumUnderPov','PctPopUnderPov','PctLess9thGrade','PctNotHSGrad'
,'PctBSorMore','PctUnemployed','PctEmploy','PctEmplManu','PctEmplProfServ','PctOccupManu','PctOccupMgmtProf','MalePctDivorce','MalePctNevMarr','FemalePctDiv','TotalPctDiv','PersPerFam','PctFam2Par','PctKids2Par','PctYoungKids2Par','PctTeen2Par','PctWorkMomYoungKids','PctWorkMom','NumIlleg','PctIlleg'
,'NumImmig','PctImmigRecent','PctImmigRec5','PctImmigRec8','PctImmigRec10','PctRecentImmig','PctRecImmig5','PctRecImmig8','PctRecImmig10','PctSpeakEnglOnly'
,'PctNotSpeakEnglWell','PctLargHouseFam','PctLargHouseOccup','PersPerOccupHous','PersPerOwnOccHous'
,'PersPerRentOccHous','PctPersOwnOccup','PctPersDenseHous','PctHousLess3BR','MedNumBR','HousVacant','PctHousOccup','PctHousOwnOcc'
,'PctVacantBoarded','PctVacMore6Mos','MedYrHousBuilt','PctHousNoPhone','PctWOFullPlumb','OwnOccLowQuart','OwnOccMedVal','OwnOccHiQuart','RentLowQ'
,'RentMedian','RentHighQ','MedRent','MedRentPctHousInc','MedOwnCostPctInc','MedOwnCostPctIncNoMtg','NumInShelters','NumStreet','PctForeignBorn'
,'PctBornSameState','PctSameHouse85','PctSameCity85','PctSameState85','LemasSwornFT','LemasSwFTPerPop','LemasSwFTFieldOps','LemasSwFTFieldPerPop'
,'LemasTotalReq','LemasTotReqPerPop','PolicReqPerOffic','PolicPerPop','RacialMatchCommPol','PctPolicWhite','PctPolicBlack','PctPolicHisp','PctPolicAsian','PctPolicMinor','OfficAssgnDrugUnits','NumKindsDrugsSeiz','PolicAveOTWorked','LandArea','PopDens','PctUsePubTrans','PolicCars','PolicOperBudg'
,'LemasPctPolicOnPatr','LemasGangUnitDeploy','LemasPctOfficDrugUn','PolicBudgPerPop'
,'ViolentCrimesPerPop']
lm_data = comm_crime #giving it a generic name so this code can be easily reused
print("Data Type", lm_data.dtypes) 
print("Shape of Data", lm_data.shape) #Dimensions of dataset
#print("Column Names", lm_data.columns) keeping this code as a comment for now – we already input the column names directly in this notebook, so there’s no need to display them again.  However, it may be useful with a different dataset, so I’m retaining this code.   
lm_data.head(5) #See top few rows of dataset

Data Type state                      int64
county                   float64
community                float64
communityname             object
fold                       int64
population               float64
householdsize            float64
racepctblack             float64
racePctWhite             float64
racePctAsian             float64
racePctHisp              float64
agePct12t21              float64
agePct12t29              float64
agePct16t24              float64
agePct65up               float64
numbUrban                float64
pctUrban                 float64
medIncome                float64
pctWWage                 float64
pctWFarmSelf             float64
pctWInvInc               float64
pctWSocSec               float64
pctWPubAsst              float64
pctWRetire               float64
medFamInc                float64
perCapInc                float64
whitePerCap              float64
blackPerCap              float64
indianPerCap             float64
AsianPerCap              float64


state  county  community        communityname  fold  population  \
0     53     NaN        NaN          Tukwilacity     1        0.00   
1     24     NaN        NaN         Aberdeentown     1        0.00   
2     34     5.0    81440.0  Willingborotownship     1        0.04   
3     42    95.0     6096.0    Bethlehemtownship     1        0.01   
4      6     NaN        NaN    SouthPasadenacity     1        0.02   

   householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.16          0.12          0.74          0.45   
1           0.42          0.49          0.56          0.17   
2           0.77          1.00          0.08          0.12   
3           0.55          0.02          0.95          0.09   
4           0.28          0.06          0.54          1.00   

          ...           LandArea  PopDens  PctUsePubTrans  PolicCars  \
0         ...               0.02     0.12            0.45        NaN   
1         ...               0.01     0.21            0.02        NaN   
2         ...               0.02     0.39            0.28        NaN   
3         ...               0.04     0.09            0.02        NaN   
4         ...               0.01     0.58            0.10        NaN   

   PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0            NaN                  NaN                  NaN   
1            NaN                  NaN                  NaN   
2            NaN                  NaN                  NaN   
3            NaN                  NaN                  NaN   
4            NaN                  NaN                  NaN   

   LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
0                  0.0              NaN                 0.67  
1                  0.0              NaN                 0.43  
2                  0.0              NaN                 0.12  
3                  0.0              NaN                 0.03  
4                  0.0              NaN                 0.14  

[5 rows x 128 columns]

In [6]:
lm_data.tail(5)

state  county  community      communityname  fold  population  \
1988     12     NaN        NaN  TempleTerracecity    10        0.01   
1989      6     NaN        NaN        Seasidecity    10        0.05   
1990      9     9.0    80070.0      Waterburytown    10        0.16   
1991     25    17.0    72600.0        Walthamcity    10        0.08   
1992      6     NaN        NaN        Ontariocity    10        0.20   

      householdsize  racepctblack  racePctWhite  racePctAsian  \
1988           0.40          0.10          0.87          0.12   
1989           0.96          0.46          0.28          0.83   
1990           0.37          0.25          0.69          0.04   
1991           0.51          0.06          0.87          0.22   
1992           0.78          0.14          0.46          0.24   

             ...           LandArea  PopDens  PctUsePubTrans  PolicCars  \
1988         ...               0.01     0.28            0.05        NaN   
1989         ...               0.02     0.37            0.20        NaN   
1990         ...               0.08     0.32            0.18       0.08   
1991         ...               0.03     0.38            0.33       0.02   
1992         ...               0.11     0.30            0.05       0.08   

      PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
1988            NaN                  NaN                  NaN   
1989            NaN                  NaN                  NaN   
1990           0.06                 0.78                  0.0   
1991           0.02                 0.79                  0.0   
1992           0.04                 0.73                  0.5   

      LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
1988                 0.00              NaN                 0.09  
1989                 0.00              NaN                 0.45  
1990                 0.91             0.28                 0.23  
1991                 0.22             0.18                 0.19  
1992                 1.00             0.13                 0.48  

[5 rows x 128 columns]

Based on this initial look, all columns except ‘community name’ are numeric.  Additionally, we see that the ‘state’, ‘county’, and ‘community’ columns are numeric, even though they effectively reflect non-numeric information.  From the attribute information, ‘fold’ is “fold number for non-random 10 fold cross validation” and not part of the predictive data.  

##Exploratory Data Analysis and Preparation

Now we’ll look at the data to see what changes (if any) need to be made to the dataset.

Now let's list the columns that have null values and how many.

In [7]:
nulls = pd.DataFrame(lm_data.isnull().sum())
nulls.columns = ['Number_of_NaNs']
nulls = nulls[nulls.Number_of_NaNs !=0]
nulls

Number_of_NaNs
county                          1173
community                       1176
OtherPerCap                        1
LemasSwornFT                    1675
LemasSwFTPerPop                 1675
LemasSwFTFieldOps               1675
LemasSwFTFieldPerPop            1675
LemasTotalReq                   1675
LemasTotReqPerPop               1675
PolicReqPerOffic                1675
PolicPerPop                     1675
RacialMatchCommPol              1675
PctPolicWhite                   1675
PctPolicBlack                   1675
PctPolicHisp                    1675
PctPolicAsian                   1675
PctPolicMinor                   1675
OfficAssgnDrugUnits             1675
NumKindsDrugsSeiz               1675
PolicAveOTWorked                1675
PolicCars                       1675
PolicOperBudg                   1675
LemasPctPolicOnPatr             1675
LemasGangUnitDeploy             1675
PolicBudgPerPop                 1675

Looking the attribute info from the webpage where the data is stored, it appears a large number of the columns have minimums of 0 and maximums of 1.  Thus, it's useful just to list those columns don't have 0 and 1 for their minimum and maximum.  Maximum values:

In [8]:
max_not1 = pd.DataFrame(lm_data.max())
max_not1.columns = ['Maximum']
max_not1 = max_not1[max_not1.Maximum !=1]
max_not1

Maximum
state                      56
county                    840
community               94597
communityname  Zanesvillecity
fold                       10

Minimum values:

In [9]:
min_not0 = pd.DataFrame(lm_data.min())
min_not0.columns = ['Minimum']
min_not0 = min_not0[min_not0.Minimum !=0]
min_not0

Minimum
state                     1
county                    1
community                70
communityname  Aberdeencity
fold                      1

Based on these minimums and maximums, it appears all numeric columns except state, county, community, and fold are normalized between 0 and 1.  

In [10]:
lm_data.describe()

state      county     community         fold   population  \
count  1993.000000  820.000000    817.000000  1993.000000  1993.000000   
mean     28.693929   58.826829  46188.336597     5.496237     0.057526   
std      16.395117  126.420560  25299.726569     2.872650     0.126903   
min       1.000000    1.000000     70.000000     1.000000     0.000000   
25%      12.000000    9.000000  25065.000000     3.000000     0.010000   
50%      34.000000   23.000000  48090.000000     5.000000     0.020000   
75%      42.000000   59.500000  66660.000000     8.000000     0.050000   
max      56.000000  840.000000  94597.000000    10.000000     1.000000   

       householdsize  racepctblack  racePctWhite  racePctAsian  racePctHisp  \
count    1993.000000   1993.000000   1993.000000   1993.000000  1993.000000   
mean        0.463462      0.179709      0.753643      0.153698     0.144009   
std         0.163731      0.253480      0.244079      0.208929     0.232549   
min         0.000000      0.000000      0.000000      0.000000     0.000000   
25%         0.350000      0.020000      0.630000      0.040000     0.010000   
50%         0.440000      0.060000      0.850000      0.070000     0.040000   
75%         0.540000      0.230000      0.940000      0.170000     0.160000   
max         1.000000      1.000000      1.000000      1.000000     1.000000   

              ...              LandArea      PopDens  PctUsePubTrans  \
count         ...           1993.000000  1993.000000     1993.000000   
mean          ...              0.065203     0.232840        0.161666   
std           ...              0.109480     0.203142        0.229111   
min           ...              0.000000     0.000000        0.000000   
25%           ...              0.020000     0.100000        0.020000   
50%           ...              0.040000     0.170000        0.070000   
75%           ...              0.070000     0.280000        0.190000   
max           ...              1.000000     1.000000        1.000000   

        PolicCars  PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
count  318.000000     318.000000           318.000000           318.000000   
mean     0.163428       0.076824             0.697956             0.440252   
std      0.215038       0.140413             0.213981             0.406434   
min      0.000000       0.000000             0.000000             0.000000   
25%      0.040000       0.020000             0.620000             0.000000   
50%      0.080000       0.030000             0.750000             0.500000   
75%      0.197500       0.060000             0.840000             1.000000   
max      1.000000       1.000000             1.000000             1.000000   

       LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
count          1993.000000       318.000000          1993.000000  
mean              0.093939         0.195252             0.237998  
std               0.240335         0.164948             0.233042  
min               0.000000         0.000000             0.000000  
25%               0.000000         0.110000             0.070000  
50%               0.000000         0.155000             0.150000  
75%               0.000000         0.220000             0.330000  
max               1.000000         1.000000             1.000000  

[8 rows x 127 columns]

Now we'll drop the columns that won't be used in the analysis. A more detailed analysis could try to analyze the data by area, but only the state data is complete here; over half the rows of county and community data are blank, so I question their usefulness.

In [11]:
lm_data_numeric = lm_data.drop(["state", "county", "community", "fold", "communityname"], axis=1)
lm_data_numeric.head(5)

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.00           0.16          0.12          0.74          0.45   
1        0.00           0.42          0.49          0.56          0.17   
2        0.04           0.77          1.00          0.08          0.12   
3        0.01           0.55          0.02          0.95          0.09   
4        0.02           0.28          0.06          0.54          1.00   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.07         0.26         0.59         0.35        0.27   
1         0.04         0.39         0.47         0.28        0.32   
2         0.10         0.51         0.50         0.34        0.21   
3         0.05         0.38         0.38         0.23        0.36   
4         0.25         0.31         0.48         0.27        0.37   

          ...           LandArea  PopDens  PctUsePubTrans  PolicCars  \
0         ...               0.02     0.12            0.45        NaN   
1         ...               0.01     0.21            0.02        NaN   
2         ...               0.02     0.39            0.28        NaN   
3         ...               0.04     0.09            0.02        NaN   
4         ...               0.01     0.58            0.10        NaN   

   PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0            NaN                  NaN                  NaN   
1            NaN                  NaN                  NaN   
2            NaN                  NaN                  NaN   
3            NaN                  NaN                  NaN   
4            NaN                  NaN                  NaN   

   LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
0                  0.0              NaN                 0.67  
1                  0.0              NaN                 0.43  
2                  0.0              NaN                 0.12  
3                  0.0              NaN                 0.03  
4                  0.0              NaN                 0.14  

[5 rows x 123 columns]

Based on analysis above, there are 22 out of 123 numeric columns where 1,675 out of 1,993 rows are null values, and one column where one value is null. For now, I will fill those null values with with the median value.  Other possibilities to address the issue are deleting rows with nulls or deleting columns with nulls, depending on the needs of the analysis.  I decided to fill all values with medians so that all rows and columns would be considered in this round of analysis.

In [12]:
lm_data_numeric_nonull = lm_data_numeric.fillna(lm_data_numeric.median())

Next is code for a correlation matrix.  A matrix wasn't required by the problem set (so I'm leaving the code as comments) but can be useful in other applications so I'm leaving it in.

In [13]:
#correlations =lm_data_numeric_nonull.corr()
#correlations

Now we’ll designate the target variable and split it off from the dataset.

In [14]:
# designate target variable name and split off from dataset
targetName = 'ViolentCrimesPerPop'
targetSeries = lm_data_numeric_nonull[targetName]
#remove target from current location and insert in column 0
del lm_data_numeric_nonull[targetName]
#lm_data_numeric_nonull.insert(0, targetName, targetSeries)
#reprint dataframe and see target is in position 0
lm_data_numeric_nonull.head(5)

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.00           0.16          0.12          0.74          0.45   
1        0.00           0.42          0.49          0.56          0.17   
2        0.04           0.77          1.00          0.08          0.12   
3        0.01           0.55          0.02          0.95          0.09   
4        0.02           0.28          0.06          0.54          1.00   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.07         0.26         0.59         0.35        0.27   
1         0.04         0.39         0.47         0.28        0.32   
2         0.10         0.51         0.50         0.34        0.21   
3         0.05         0.38         0.38         0.23        0.36   
4         0.25         0.31         0.48         0.27        0.37   

        ...         PolicAveOTWorked  LandArea  PopDens  PctUsePubTrans  \
0       ...                     0.26      0.02     0.12            0.45   
1       ...                     0.26      0.01     0.21            0.02   
2       ...                     0.26      0.02     0.39            0.28   
3       ...                     0.26      0.04     0.09            0.02   
4       ...                     0.26      0.01     0.58            0.10   

   PolicCars  PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0       0.08           0.03                 0.75                  0.5   
1       0.08           0.03                 0.75                  0.5   
2       0.08           0.03                 0.75                  0.5   
3       0.08           0.03                 0.75                  0.5   
4       0.08           0.03                 0.75                  0.5   

   LemasPctOfficDrugUn  PolicBudgPerPop  
0                  0.0            0.155  
1                  0.0            0.155  
2                  0.0            0.155  
3                  0.0            0.155  
4                  0.0            0.155  

[5 rows x 122 columns]

Now we’re going to split the data into training and test sets, in a 75/25 ratio.  

In [15]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(lm_data_numeric_nonull, targetSeries, test_size=0.25, random_state=33)

The following is code to look at min, max, and mean of the training and test sets.  We won't execute it here, but I'm leaving it in for future use.

In [16]:
#print(np.max(X_train), np.min(X_train), np.mean(X_train), np.max(y_train), np.min(y_train), np.mean(y_train))

###Normalize data

Normally, we would normalize the data before proceeding, but we've already verified that all values are between zero and one.  Thus, I'm leaving this code in but as a comment so it can be used in future projects.

In [17]:
#from sklearn.preprocessing import StandardScaler
#X_train = StandardScaler().fit_transform(X_train)
#scalerX = StandardScaler().fit(X_train)
#scalery = StandardScaler().fit(y_train)

#X_train = scalerX.transform(X_train)
#y_train = StandardScaler().fit_transform((y_train)
#X_test = StandardScaler().fit_transform(X_test)
#y_test = StandardScaler().fit_transform(y_test)

#print(np.max(X_train), np.min(X_train), np.mean(X_train), np.max(y_train), np.min(y_train), np.mean(y_train))


##Running linear models

Now that the dataset is prepared, we can run the different linear models.

###Ridge Regression

The first linear model we'll use is the Ridge Regression.  As the Ridge has a peanlty incorporated (alpha), we'll start with the default value and then use a grid search to find the optimum value.  In the case of the Ridge Regression, it uses 'L2' for its penalty (alpha), which penalizes based on the square of the weight.  It tends to result in fewer factors reduced to zero than 'L1', which uses the absolute value of the weight.

In [18]:
# Ridge Regression
from sklearn.linear_model import Ridge
model_rg = Ridge() #ridge with default alpha
model_rg.fit(X_train, y_train)
print(model_rg) #shows parameters and their default values
expected_rg = y_train
predicted_rg = model_rg.predict(X_train)
print("Ridge Model Coefficients:")
print(model_rg.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_rg,predicted_rg))
print("R-Squared:")
print(r2_score(expected_rg,predicted_rg))
print("Explained Variance Score:")
print(explained_variance_score(expected_rg,predicted_rg))
print("Median Absolute Error:") 
print(median_absolute_error(expected_rg,predicted_rg))

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Ridge Model Coefficients:
[-0.05689025  0.01982886  0.17738657 -0.02907009 -0.00728552  0.06039156
  0.04756712 -0.18240027 -0.05822169  0.04049969 -0.0457667   0.04888704
  0.05954256 -0.14380607  0.01622719 -0.15766169  0.0310556   0.03186793
 -0.08981279  0.02736342 -0.00842146 -0.11071426 -0.02102865 -0.02487761
  0.0067258   0.02594549  0.02752024  0.02980123 -0.13466184 -0.07266441
  0.06336806  0.02802515 -0.00594524  0.15017501 -0.0344746   0.02036406
  0.02830624  0.02041756  0.18829914  0.15561227 -0.12692222 -0.01877988
  0.05777633 -0.0432426  -0.14786919 -0.05135192  0.00935541  0.02841277
 -0.14902372 -0.10787731  0.14831198 -0.03420198  0.03740341 -0.0171951
  0.01251498 -0.01172876 -0.05864224 -0.02587833  0.07553637  0.02281385
  0.03412834 -0.12839741 -0.0622747  -0.04606317  0.16059874 -0.07796642
 -0.03058179 -0.12310599  0.1142622   0.07

With the default value of alpha = 1, the mean squared error in this case was 0.0162, the median absolute error 0.0629, and the R-squared was 0.693.  The default settings for the alpha didn't reduce any of the factors to zero, so this model has 122 variables, which strikes me as very complex.  Now let's do the grid search and see what the optimum value of alpha is.

In [19]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
param_grid = {"alpha": [.01,.05,.1,.5, 1, 2,3,4,5,6,7,8,9,10,15,20]}
# run grid search
grid_search = GridSearchCV(model_rg, param_grid=param_grid,n_jobs=-1,cv=5,return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best", grid_search.best_params_) 
print("Grid Scores:")
pd.DataFrame(grid_search.cv_results_)

Best {'alpha': 4}
Grid Scores:


mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
0        0.156250         0.003125         0.618339          0.710268   
1        0.087498         0.003125         0.624660          0.709119   
2        0.125000         0.000000         0.628079          0.707904   
3        0.187500         0.000000         0.635553          0.702032   
4        0.206251         0.003125         0.637749          0.697569   
5        0.100002         0.000000         0.639053          0.691644   
6        0.071874         0.000000         0.639402          0.687542   
7        0.053124         0.000000         0.639466          0.684390   
8        0.049998         0.009375         0.639406          0.681829   
9        0.090623         0.000000         0.639280          0.679670   
10       0.075001         0.000000         0.639117          0.677803   
11       0.103124         0.003124         0.638930          0.676158   
12       0.090624         0.000000         0.638728          0.674686   
13       0.074998         0.000000         0.638515          0.673353   
14       0.084375         0.000000         0.637372          0.668072   
15       0.053125         0.000000         0.636174          0.664143   

   param_alpha           params  rank_test_score  split0_test_score  \
0         0.01  {'alpha': 0.01}               16           0.659092   
1         0.05  {'alpha': 0.05}               15           0.666363   
2          0.1   {'alpha': 0.1}               14           0.670788   
3          0.5   {'alpha': 0.5}               13           0.680854   
4            1     {'alpha': 1}               10           0.683252   
5            2     {'alpha': 2}                6           0.684390   
6            3     {'alpha': 3}                3           0.684732   
7            4     {'alpha': 4}                1           0.684883   
8            5     {'alpha': 5}                2           0.684931   
9            6     {'alpha': 6}                4           0.684904   
10           7     {'alpha': 7}                5           0.684818   
11           8     {'alpha': 8}                7           0.684682   
12           9     {'alpha': 9}                8           0.684507   
13          10    {'alpha': 10}                9           0.684298   
14          15    {'alpha': 15}               11           0.682915   
15          20    {'alpha': 20}               12           0.681216   

    split0_train_score  split1_test_score       ...         split2_test_score  \
0             0.699544           0.554854       ...                  0.669269   
1             0.698432           0.562578       ...                  0.673583   
2             0.697234           0.567267       ...                  0.675323   
3             0.691048           0.578250       ...                  0.680663   
4             0.686348           0.581272       ...                  0.684050   
5             0.680288           0.581993       ...                  0.687805   
6             0.676162           0.581244       ...                  0.689880   
7             0.672997           0.580255       ...                  0.691192   
8             0.670415           0.579275       ...                  0.692087   
9             0.668227           0.578361       ...                  0.692728   
10            0.666324           0.577523       ...                  0.693203   
11            0.664638           0.576756       ...                  0.693562   
12            0.663122           0.576053       ...                  0.693835   
13            0.661746           0.575406       ...                  0.694044   
14            0.656258           0.572770       ...                  0.694484   
15            0.652164           0.570757       ...                  0.694358   

    split2_train_score  split3_test_score  split3_train_score  \
0             0.701245           0.585374            0.714341   
1             0.699904           0.590293     

The grid search data computes alpha=4 as the best result.  Let's put that into the Ridge model and see the result.

In [20]:
# Ridge Regression
from sklearn.linear_model import Ridge
model_rg_al = Ridge(alpha=4)
model_rg_al.fit(X_train, y_train)
print(model_rg_al) #shows parameters and their default values
expected_rg = y_train
predicted_rg = model_rg_al.predict(X_train)
print("Ridge Model Coefficients:")
print(model_rg_al.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_rg,predicted_rg))
print("R-Squared:")
print(r2_score(expected_rg,predicted_rg))
print("Explained Variance Score:")
print(explained_variance_score(expected_rg,predicted_rg))
print("Median Absolute Error:") 
print(median_absolute_error(expected_rg,predicted_rg))

Ridge(alpha=4, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)
Ridge Model Coefficients:
[ -2.58038849e-02   1.73957827e-02   1.52467814e-01  -5.77936567e-02
  -1.08103032e-02   2.46478713e-02   4.71905543e-03  -8.29459282e-02
  -3.05349647e-02   2.83395358e-02  -2.31135627e-02   4.58867617e-02
   3.15990441e-02  -7.40994264e-02   5.81672884e-03  -1.03077639e-01
   2.12930393e-02   3.26001524e-02  -7.37291954e-02   5.25248089e-03
  -1.19448400e-02  -3.59773883e-02  -1.74053528e-02  -2.34060458e-02
   8.89655504e-03   2.46861609e-02   2.33785807e-02  -4.84607878e-03
  -8.03460134e-02  -4.10775738e-02   3.51257166e-02  -3.41729603e-03
  -1.61930263e-02   5.94626354e-02  -2.67095942e-02   1.40832500e-02
   1.61750144e-02   1.44337285e-03   1.13565013e-01   7.80467233e-02
  -4.61848734e-02   1.37669205e-02   3.58270654e-02  -5.40249457e-02
  -1.00715775e-01  -6.12182759e-02  -2.38302047e-03  -3.50393837e-04
  -9.55703011e-02 

The resulting Ridge model has a mean squared error of 0.0168, a median absolute error of 0.0625, and the R-squared was 0.682, which is slightly less accurate than the default setting of alpha=1.  Again, the alpha didn't reduce any of the factors to zero, so this model still has 122 variables. 

###LASSO Regression Model

Now we'll try the LASSO Regression, which like the Ridge has a penalty incorporated (alpha), so we'll start with the default value and then use a grid search to find the optimum value.  Unlike the Ridge Regression, it uses 'L1' for its penalty (alpha), which penalizes based on the absolute value of the weight, typically resulting in more factors being reduced to zero.  Thus, it can be useful to reduce a large number of dimensions.

In [21]:
# Lasso Regression
# fit a LASSO model to the data
from sklearn.linear_model import Lasso
model_ls = Lasso() #LASSO with default vaues
model_ls.fit(X_train, y_train)
print(model_ls) #shows model parameters
# make predictions
expected_ls = y_train
predicted_ls = model_ls.predict(X_train)
# summarize the fit of the model
print("LASSO Model Coefficients:")
print(model_ls.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_ls,predicted_ls))
print("R-Squared:")
print(r2_score(expected_ls,predicted_ls))
print("Explained Variance Score:")
print(explained_variance_score(expected_ls,predicted_ls))
print("Median Absolute Error:") 
print(median_absolute_error(expected_ls,predicted_ls))

Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
LASSO Model Coefficients:
[ 0. -0.  0. -0.  0.  0.  0.  0.  0.  0.  0.  0. -0. -0. -0. -0.  0.  0.
 -0. -0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0. -0.  0. -0. -0. -0.
  0. -0.  0.  0.  0.  0.  0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0.  0. -0.  0.  0.  0. -0. -0.  0. -0.  0.  0. -0.  0.
 -0. -0.  0.  0. -0.  0.  0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.
  0.  0. -0. -0.  0. -0.  0.  0. -0.  0.  0.  0.  0.  0. -0. -0.  0.  0.
  0.  0.  0.  0.  0.  0.  0.  0.  0.  0. -0. -0.  0.  0.]
Mean Squared Error:
0.0528797571365
R-Squared:
0.0
Explained Variance Score:
1.11022302463e-16
Median Absolute Error:
0.156211512718


The default setting for the alpha (1) reduced all of the factors to zero!  The mean squared error in this case was 0.0528 and the R-squared was 0.0, but those don't have any effective meaning here.   Now let's do the grid search and see what the optimum value of alpha is.

In [22]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
param_grid = {"alpha": [.000001, .00001, .0001,.001,.01,.02, .05, .1, .2, .3,.4,.5,.6,.7,.8,.9,1]}
# run grid search
grid_search = GridSearchCV(model_ls, param_grid=param_grid,n_jobs=-1,cv=5,return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best", grid_search.best_params_) 
print("Grid Scores:")
pd.DataFrame(grid_search.cv_results_)

Best {'alpha': 0.0001}
Grid Scores:


mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
0        1.012501         0.003124         0.614824          0.710420   
1        1.150004         0.000000         0.623573          0.709382   
2        0.490623         0.006250         0.640213          0.695387   
3        0.106250         0.000000         0.636720          0.656254   
4        0.081252         0.000000         0.541757          0.551396   
5        0.071879         0.000000         0.397880          0.408206   
6        0.046873         0.003126        -0.011480          0.000000   
7        0.037500         0.000000        -0.011480          0.000000   
8        0.031252         0.000000        -0.011480          0.000000   
9        0.037502         0.000000        -0.011480          0.000000   
10       0.046875         0.000000        -0.011480          0.000000   
11       0.050001         0.003124        -0.011480          0.000000   
12       0.028123         0.003126        -0.011480          0.000000   
13       0.049998         0.000000        -0.011480          0.000000   
14       0.040625         0.000000        -0.011480          0.000000   
15       0.024999         0.000000        -0.011480          0.000000   
16       0.037501         0.000000        -0.011480          0.000000   

   param_alpha             params  rank_test_score  split0_test_score  \
0        1e-06   {'alpha': 1e-06}                4           0.654662   
1        1e-05   {'alpha': 1e-05}                3           0.665374   
2       0.0001  {'alpha': 0.0001}                1           0.687723   
3        0.001   {'alpha': 0.001}                2           0.683080   
4         0.01    {'alpha': 0.01}                5           0.575081   
5         0.02    {'alpha': 0.02}                6           0.411280   
6         0.05    {'alpha': 0.05}                7          -0.012055   
7          0.1     {'alpha': 0.1}                7          -0.012055   
8          0.2     {'alpha': 0.2}                7          -0.012055   
9          0.3     {'alpha': 0.3}                7          -0.012055   
10         0.4     {'alpha': 0.4}                7          -0.012055   
11         0.5     {'alpha': 0.5}                7          -0.012055   
12         0.6     {'alpha': 0.6}                7          -0.012055   
13         0.7     {'alpha': 0.7}                7          -0.012055   
14         0.8     {'alpha': 0.8}                7          -0.012055   
15         0.9     {'alpha': 0.9}                7          -0.012055   
16           1       {'alpha': 1}                7          -0.012055   

    split0_train_score  split1_test_score       ...         split2_test_score  \
0             0.699724           0.552699       ...                  0.665006   
1             0.698592           0.559881       ...                  0.675518   
2             0.683821           0.583207       ...                  0.684776   
3             0.644921           0.571983       ...                  0.693835   
4             0.540988           0.493454       ...                  0.590903   
5             0.389954           0.373167       ...                  0.419558   
6             0.000000          -0.030598       ...                 -0.000357   
7             0.000000          -0.030598       ...                 -0.000357   
8             0.000000          -0.030598       ...                 -0.000357   
9             0.000000          -0.030598       ...                 -0.000357   
10            0.000000          -0.030598       ...                 -0.000357   
11            0.000000          -0.030598       ...                 -0.000357   
12            0.000000          -0.030598       ...                 -0.000357   
13            0.000000          -0.030598       ...                 -0.000357   
14            0.000000          -0.030598       ...                 -0.000357   
15            0.000000          -0.030598       ...                 -0.000357 

I had to use much lower values of alpha in this case when hunting for the optimal solution (not surprisingly, as we saw an alpha of one reducing all terms to zero).  The grid search determined that alpha=0.0001 was best.  Let's put that in the LASSO model and see the result.

In [23]:
# Lasso Regression with alpha value determined by grid search
# fit a LASSO model to the data
from sklearn.linear_model import Lasso
model_ls_al = Lasso(alpha=0.0001) 
model_ls_al.fit(X_train, y_train)
print(model_ls_al) #shows model parameters
# make predictions
expected_ls = y_train
predicted_ls = model_ls_al.predict(X_train)
# summarize the fit of the model
print("LASSO Model Coefficients:")
print(model_ls_al.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_ls,predicted_ls))
print("R-Squared:")
print(r2_score(expected_ls,predicted_ls))
print("Explained Variance Score:")
print(explained_variance_score(expected_ls,predicted_ls))
print("Median Absolute Error:") 
print(median_absolute_error(expected_ls,predicted_ls))

Lasso(alpha=0.0001, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)
LASSO Model Coefficients:
[-0.02141139  0.          0.19437811 -0.00251264 -0.          0.03975464
  0.029732   -0.24149453  0.          0.         -0.          0.04549541
  0.         -0.19320203  0.01064556 -0.14588531  0.          0.00461044
 -0.09377493  0.         -0.         -0.04255989 -0.01150573 -0.02183552
  0.00715895  0.02277153  0.02176504 -0.         -0.13365505 -0.02033469
  0.          0.         -0.          0.13384331 -0.01702572  0.02229886
  0.          0.          0.19469328  0.16304113 -0.12163629 -0.          0.
 -0.         -0.186444   -0.04476262  0.          0.00694928 -0.12855144
 -0.1016484   0.16732954 -0.          0.01766404  0.          0.          0.
 -0.         -0.          0.          0.          0.         -0.11028525
 -0.06793987 -0.          0.16394893 -0.03

The resulting LASSO model has a mean squared error of 0.0164, a median absolute error of 0.0629, and a R-squared of 0.6892, on par with the Ridge results.  Slightly less than half of the terms were penalized to zero.

###Elastic Net Regression

Now we'll try the Elasitc Net Regression, which also has a penalty incorporated (alpha), so we'll start with the default value and then use a grid search to find the optimum value.  The Elasitc Net uses a combination of L1 and L2 for its penalty. 

In [24]:
# ElasticNet Regression
# fit a model to the data
from sklearn.linear_model import ElasticNet
model_en = ElasticNet()
model_en.fit(X_train, y_train)
print(model_en)
# make predictions
expected_en = y_train
predicted_en = model_en.predict(X_train)
# summarize the fit of the model
print("Elastic Net Model Coefficients:")
print(model_en.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_en,predicted_en))
print("R-Squared:")
print(r2_score(expected_en,predicted_en))
print("Explained Variance Score:")
print(explained_variance_score(expected_en,predicted_en))
print("Median Absolute Error:") 
print(median_absolute_error(expected_en,predicted_en))

ElasticNet(alpha=1.0, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Elastic Net Model Coefficients:
[ 0. -0.  0. -0.  0.  0.  0.  0.  0.  0.  0.  0. -0. -0. -0. -0.  0.  0.
 -0. -0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0. -0.  0. -0. -0. -0.
  0. -0.  0.  0.  0.  0.  0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0.  0. -0.  0.  0.  0. -0. -0.  0. -0.  0.  0. -0.  0.
 -0. -0.  0.  0. -0.  0.  0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.
  0.  0. -0. -0.  0. -0.  0.  0. -0.  0.  0.  0.  0.  0. -0. -0.  0.  0.
  0.  0.  0.  0.  0.  0.  0.  0.  0.  0. -0. -0.  0.  0.]
Mean Squared Error:
0.0528797571365
R-Squared:
0.0
Explained Variance Score:
1.11022302463e-16
Median Absolute Error:
0.156211512718


Similar to the LASSO, the default setting for the alpha (1) reduced all of the factors to zero!  The mean squared error in this case was 0.0528 and the R-squared was 0.0, but those don't have any effective meaning here.   Now let's do the grid search and see what the optimum value of alpha is.

In [25]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
param_grid = {"alpha": [.000001, .00001, .0001,.001,.01,.02, .05, .1, .2, .3,.4,.5,.6,.7,.8,.9,1]}
# run grid search
grid_search = GridSearchCV(model_en, param_grid=param_grid,n_jobs=-1,cv=5,return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best", grid_search.best_params_) 
print("Grid Scores:")
pd.DataFrame(grid_search.cv_results_)

Best {'alpha': 0.001}
Grid Scores:


mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
0        1.153124         0.006249         0.614601          0.710425   
1        1.531252         0.000000         0.621469          0.709861   
2        0.765627         0.012500         0.635886          0.701350   
3        0.203126         0.009374         0.640361          0.666325   
4        0.118752         0.009374         0.593667          0.606021   
5        0.056251         0.003125         0.526964          0.536701   
6        0.103127         0.015625         0.264351          0.276071   
7        0.056250         0.003125        -0.011480          0.000000   
8        0.056253         0.009374        -0.011480          0.000000   
9        0.046874         0.000000        -0.011480          0.000000   
10       0.049999         0.003125        -0.011480          0.000000   
11       0.059375         0.003125        -0.011480          0.000000   
12       0.068752         0.000000        -0.011480          0.000000   
13       0.031249         0.003125        -0.011480          0.000000   
14       0.053124         0.003125        -0.011480          0.000000   
15       0.037502         0.006249        -0.011480          0.000000   
16       0.050001         0.000000        -0.011480          0.000000   

   param_alpha             params  rank_test_score  split0_test_score  \
0        1e-06   {'alpha': 1e-06}                4           0.654428   
1        1e-05   {'alpha': 1e-05}                3           0.662647   
2       0.0001  {'alpha': 0.0001}                2           0.682017   
3        0.001   {'alpha': 0.001}                1           0.686815   
4         0.01    {'alpha': 0.01}                5           0.628332   
5         0.02    {'alpha': 0.02}                6           0.557824   
6         0.05    {'alpha': 0.05}                7           0.260377   
7          0.1     {'alpha': 0.1}                8          -0.012055   
8          0.2     {'alpha': 0.2}                8          -0.012055   
9          0.3     {'alpha': 0.3}                8          -0.012055   
10         0.4     {'alpha': 0.4}                8          -0.012055   
11         0.5     {'alpha': 0.5}                8          -0.012055   
12         0.6     {'alpha': 0.6}                8          -0.012055   
13         0.7     {'alpha': 0.7}                8          -0.012055   
14         0.8     {'alpha': 0.8}                8          -0.012055   
15         0.9     {'alpha': 0.9}                8          -0.012055   
16           1       {'alpha': 1}                8          -0.012055   

    split0_train_score  split1_test_score       ...         split2_test_score  \
0             0.699731           0.552433       ...                  0.664972   
1             0.699087           0.557980       ...                  0.673166   
2             0.690390           0.576770       ...                  0.679929   
3             0.655068           0.578647       ...                  0.697370   
4             0.593174           0.538016       ...                  0.652164   
5             0.526477           0.483746       ...                  0.572517   
6             0.253652           0.256325       ...                  0.274877   
7             0.000000          -0.030598       ...                 -0.000357   
8             0.000000          -0.030598       ...                 -0.000357   
9             0.000000          -0.030598       ...                 -0.000357   
10            0.000000          -0.030598       ...                 -0.000357   
11            0.000000          -0.030598       ...                 -0.000357   
12            0.000000          -0.030598       ...                 -0.000357   
13            0.000000          -0.030598       ...                 -0.000357   
14            0.000000          -0.030598       ...                 -0.000357   
15            0.000000          -0.030598       ...                 -0.000357 

Similar to LASSO, I had to use much lower values of alpha in this case when hunting for the optimal solution (not surprisingly, as we saw an alpha of one reducing all terms to zero).  The grid search determined that alpha=0.001 was best.  Let's put that in the Elastic Net model and see the result.

In [26]:
# ElasticNet Regression
# fit a model to the data
from sklearn.linear_model import ElasticNet
model_en_al = ElasticNet(alpha=0.001)
model_en_al.fit(X_train, y_train)
print(model_en_al)
# make predictions
expected_en = y_train
predicted_en = model_en_al.predict(X_train)
# summarize the fit of the model
print("Elastic Net Model Coefficients:")
print(model_en_al.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_en,predicted_en))
print("R-Squared:")
print(r2_score(expected_en,predicted_en))
print("Explained Variance Score:")
print(explained_variance_score(expected_en,predicted_en))
print("Median Absolute Error:") 
print(median_absolute_error(expected_en,predicted_en))

ElasticNet(alpha=0.001, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=None, selection='cyclic', tol=0.0001, warm_start=False)
Elastic Net Model Coefficients:
[ 0.         -0.          0.16930543 -0.02464761  0.          0.         -0.
 -0.02753984 -0.          0.          0.          0.04142679  0.
 -0.04425236 -0.         -0.05184516  0.          0.00399158 -0.01894466
 -0.          0.          0.         -0.         -0.00318068  0.
  0.01190197  0.0004996   0.         -0.         -0.          0.         -0.
 -0.          0.         -0.         -0.         -0.         -0.
  0.14538977  0.          0.          0.          0.         -0.
 -0.15708745 -0.03874272 -0.         -0.         -0.0716888   0.
  0.19775075  0.          0.          0.          0.          0.          0.
  0.          0.          0.          0.         -0.          0.          0.
  0.         -0.          0.         -0.    

The resulting Elastic Net model has a mean squared error of 0.0178, a median absolute error of 0.0629, and a R-squared of 0.6636, slightly worse than the Ridge and LASSO results.  Roughly two thirds of the terms were penalized to zero.

###LARS (Least Angle Regression) model

Now we'll look at the LARS, which conducts the regression in steps by finding the column most closely correlated with the target. 

In [27]:
#LARS Regression Model- Least Angle Regression model
from sklearn import linear_model
model_LAR = linear_model.Lars() #default numer of n_nonzero_coefs
model_LAR.fit(X_train, y_train)
print(model_LAR)
# make predictions
expected_LAR = y_train
predicted_LAR = model_LAR.predict(X_train)
# summarize the fit of the model
#mse_LAR = np.mean((predicted_LAR-expected_LAR)**2)
print("LARS intercept:")
print(model_LAR.intercept_)
print("LARS Coefficients:")
print(model_LAR.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_LAR,predicted_LAR))
print("R-Squared:")
print(r2_score(expected_LAR,predicted_LAR))
print("Explained Variance Score:")
print(explained_variance_score(expected_LAR,predicted_LAR))
print("Median Absolute Error:") 
print(median_absolute_error(expected_LAR,predicted_LAR))

Lars(copy_X=True, eps=2.2204460492503131e-16, fit_intercept=True,
   fit_path=True, n_nonzero_coefs=500, normalize=True, positive=False,
   precompute='auto', verbose=False)
LARS intercept:
-60365375308.6
LARS Coefficients:
[ -1.51547539e+10   3.99384787e+08  -1.67311467e+08  -1.60193635e+09
  -5.65357082e+07   4.18802181e+08  -5.66050119e+08   6.01058130e+09
  -2.94044413e+09   1.78496875e+09   1.02373212e+10  -2.68445469e+08
   6.56624243e+09  -6.44944761e+08  -3.13894156e+08   5.00881909e+08
   7.60009012e+08  -1.00114441e+09  -1.40889266e+09  -3.92735020e+09
  -1.28361052e+08  -3.42721197e+09  -1.37220316e+08   8.45317669e+07
  -1.24176636e+07   2.19354643e+08   2.44080425e+08   1.31773268e+10
   1.54430923e+09   7.18437934e+08  -1.18718395e+09   1.41875756e+09
   1.15101585e+08  -2.39918787e+08  -3.00509641e+08  -3.82746040e+08
  -2.36403050e+08   1.18209484e+08   1.70957864e+10  -1.23677803e+09
   2.17755723e+10  -3.53178493e+10  -7.35986773e+09   2.95252974e+09
  -3.50678626e+09

The error numbers are so large here that the model is effectively meaningless.  The default number of remaining values (the paremeter is called "n nonzero coefficients") is 500 in the LARS, so let's use a grid serach to find the optimal number.

In [28]:
# use a full grid over several parameters and cross validate 5 times
from sklearn.model_selection import GridSearchCV
param_grid = {"n_nonzero_coefs": [1,2,3,4,5,6,7,8,9,10,20,30,40,50,60,70,80,90,100,110,120]}
#param_grid={"alpha": [0,10,1]} #this does a range 1 through 10 changes by a factor of 1. 
#param_grid={"alpha": [.01,1,.05]} #this does a range 1 through 1 changes by a factor of .05

# run grid search
grid_search = GridSearchCV(model_LAR, param_grid=param_grid,n_jobs=-1,cv=5,return_train_score=True)
grid_search.fit(X_train, y_train)
print("Best", grid_search.best_params_) 
print("Grid Scores:")
pd.DataFrame(grid_search.cv_results_)

Best {'n_nonzero_coefs': 8}
Grid Scores:


mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
0        0.159376         0.000000         0.042458          0.055538   
1        0.087499         0.000000         0.368916          0.380232   
2        0.100002         0.003125         0.533453          0.542660   
3        0.121875         0.000000         0.554618          0.565299   
4        0.137501         0.000000         0.569939          0.582132   
5        0.118749         0.006250         0.580946          0.592724   
6        0.125001         0.003124         0.588785          0.600610   
7        0.125000         0.003126         0.591949          0.603523   
8        0.118750         0.003125         0.575592          0.587252   
9        0.121877         0.000000         0.557389          0.547823   
10       0.146873         0.000000         0.229753          0.063721   
11       0.165627         0.003125        -0.130660         -0.444349   
12       0.165623         0.006251        -1.369488         -2.253636   
13       0.206250         0.000000       -19.284316        -24.974841   
14       0.225001         0.003125       -31.571999        -40.230665   
15       0.190624         0.000000       -73.836959        -88.003883   
16       0.193749         0.000000      -145.024743       -163.954110   
17       0.193748         0.006250      -230.293171       -250.394387   
18       0.334375         0.000000      -302.039280       -320.288884   
19       0.493751         0.018750      -477.007306       -474.877894   
20       0.381250         0.003126      -595.509479       -576.033618   

   param_n_nonzero_coefs                    params  rank_test_score  \
0                      1    {'n_nonzero_coefs': 1}               11   
1                      2    {'n_nonzero_coefs': 2}                9   
2                      3    {'n_nonzero_coefs': 3}                8   
3                      4    {'n_nonzero_coefs': 4}                7   
4                      5    {'n_nonzero_coefs': 5}                5   
5                      6    {'n_nonzero_coefs': 6}                3   
6                      7    {'n_nonzero_coefs': 7}                2   
7                      8    {'n_nonzero_coefs': 8}                1   
8                      9    {'n_nonzero_coefs': 9}                4   
9                     10   {'n_nonzero_coefs': 10}                6   
10                    20   {'n_nonzero_coefs': 20}               10   
11                    30   {'n_nonzero_coefs': 30}               12   
12                    40   {'n_nonzero_coefs': 40}               13   
13                    50   {'n_nonzero_coefs': 50}               14   
14                    60   {'n_nonzero_coefs': 60}               15   
15                    70   {'n_nonzero_coefs': 70}               16   
16                    80   {'n_nonzero_coefs': 80}               17   
17                    90   {'n_nonzero_coefs': 90}               18   
18                   100  {'n_nonzero_coefs': 100}               19   
19                   110  {'n_nonzero_coefs': 110}               20   
20                   120  {'n_nonzero_coefs': 120}               21   

    split0_test_score  split0_train_score  split1_test_score       ...         \
0            0.015237            0.024545           0.080367       ...          
1            0.358786            0.332586           0.337893       ...          
2            0.573866            0.533793           0.475356       ...          
3            0.599488            0.556718           0.501851       ...          
4            0.611154            0.567458           0.518108       ...          
5            0.611595            0.567857           0.523949       ...          
6            0.633290            0.587056           0.524005       ...          
7            0.633866            0.587424           0.532016       ...          
8            0.634124            0.587610           0.533272       ...          
9            0.631648   

The grid serach determined that 8 nonzero coefficients is the best answer.  Let's put that number into the LARS.

In [29]:
#LARS Regression Model- Least Angle Regression model
from sklearn import linear_model
model_LAR_cf = linear_model.Lars(n_nonzero_coefs=8)
model_LAR_cf.fit(X_train, y_train)
print(model_LAR_cf)
# make predictions
expected_LAR = y_train
predicted_LAR = model_LAR_cf.predict(X_train)
# summarize the fit of the model
print("LARS intercept:")
print(model_LAR_cf.intercept_)
print("LARS Coefficients:")
print(model_LAR_cf.coef_)
print("Mean Squared Error:")
print(mean_squared_error(expected_LAR,predicted_LAR))
print("R-Squared:")
print(r2_score(expected_LAR,predicted_LAR))
print("Explained Variance Score:")
print(explained_variance_score(expected_LAR,predicted_LAR))
print("Median Absolute Error:") 
print(median_absolute_error(expected_LAR,predicted_LAR))

Lars(copy_X=True, eps=2.2204460492503131e-16, fit_intercept=True,
   fit_path=True, n_nonzero_coefs=8, normalize=True, positive=False,
   precompute='auto', verbose=False)
LARS intercept:
0.498311886923
LARS Coefficients:
[  0.00000000e+00   0.00000000e+00   0.00000000e+00  -1.58651432e-01
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00
   0.00000000e+00   0.00000000e+00   7.64547341e-02   0.00000000e+00
   0.00000000e+00  -4.29670793e-02   0.00000000e+00   0.00000000e+00
  -3.32715597e-01  

The LARS has a mean squared error of 0.0216, a median absolute error of 0.0720, and an R-squared of 0.591, slightly worse than the other results seen so far.    

###Cross validation

Now we're going to conduct a cross validation on the Ridge model with the alpha=4 as found during the grid search.    Cross validation is a process where we take a subset of the data (in this case, 10% of it) and a this data is compared to a model generated by the remaining data (in this case, 90% of it).  This process is repeated for a total of 10 times (in our example), with each run using a different 10% of the data.  

In [30]:
#verify Ridge with Cross Validation
scores = cross_val_score(model_rg_al, X_train, y_train, cv=10)
print("Cross Validation Score for each K",scores)
print("Mean Cross Validation Score", scores.mean()) 

Cross Validation Score for each K [ 0.71818732  0.64075015  0.63274775  0.49837093  0.72692664  0.64143831
  0.62025856  0.60044992  0.63444518  0.63891038]
Mean Cross Validation Score 0.635248514688


The cross validation revealed a wide range of scores from 0.498 to 0.727.  This suggests the model is overfit, which not entirely surprising given that it contains 122 variables.  Since the LASSO results were similar, we'll run a Cross Validation on the LASSO as well:

In [31]:
#verify LASSO with Cross Validation
scores = cross_val_score(model_ls_al, X_train, y_train, cv=10)
print("Cross Validation Score for each K",scores)
print("Mean Cross Validation Score", scores.mean())

Cross Validation Score for each K [ 0.7193922   0.64335571  0.63249457  0.50557729  0.71939543  0.63940197
  0.63024936  0.60924868  0.63448746  0.63162436]
Mean Cross Validation Score 0.636522702073


The LASSO cross validation results also show a significant spread, from 0.506 to 0.719, but not quite as large as the Ridge.

##Conclusion

Based on the Mean Squared Error, Median Absolute Error, and R Squared scores, I judge the Ridge  and LASSO models to be nearly tied in performance.  That said, I would choose the LASSO as the cross validation results for the Ridge were worse, thus the Ridge model is likely more overfit than the LASSO.  Follow-on analysis should look at which variables should be eliminated from the model to achieve less variance, and other methods to achieve more consistant results.  